### ISSUE TITLE
mix compile --all-warnings returns crash

### ISSUE THREAD

**Issue by mojidabckuu** [elixir-lang/elixir#10094]
### Precheck 

Possibly related https://github.com/elixir-lang/elixir/issues/10073? And may be fixed. As I can see there is a test added for `--warnings-as-errors` with `catch_exit` but for `--all-warnings` there is no the same assert.

### Environment

Elixir 1.10.3 (compiled with Erlang/OTP 22
macOS 10.15.4

### Current behavior

```
mix compile --all-warnings

** (FunctionClauseError) no function clause matching in :elixir_errors.erl_warn/3    
    
    The following arguments were given to :elixir_errors.erl_warn/3:
    
        # 1
        nil
    
        # 2
        "/Users/vladgorbenko/Documents/elixir/myapp/lib/services/my_supervisor.ex"
    
        # 3
        ["Supervisor.Spec.supervise/2", " is deprecated. ", "Use the new child specifications outlined in the Supervisor module instead"]
    
    (elixir 1.10.3) src/elixir_errors.erl:13: :elixir_errors.erl_warn/3
    (mix 1.10.3) lib/mix/compilers/elixir.ex:518: anonymous fn/3 in Mix.Compilers.Elixir.show_warnings/1
    (elixir 1.10.3) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (mix 1.10.3) lib/mix/compilers/elixir.ex:517: anonymous fn/2 in Mix.Compilers.Elixir.show_warnings/1
    (elixir 1.10.3) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (mix 1.10.3) lib/mix/compilers/elixir.ex:514: Mix.Compilers.Elixir.show_warnings/1
    (mix 1.10.3) lib/mix/compilers/elixir.ex:110: Mix.Compilers.Elixir.compile/6
    (mix 1.10.3) lib/mix/task.ex:330: Mix.Task.run_task/3
    (mix 1.10.3) lib/mix/tasks/compile.all.ex:76: Mix.Tasks.Compile.All.run_compiler/2
    (mix 1.10.3) lib/mix/tasks/compile.all.ex:56: Mix.Tasks.Compile.All.do_compile/4
    (mix 1.10.3) lib/mix/tasks/compile.all.ex:27: anonymous fn/2 in Mix.Tasks.Compile.All.run/1
    (mix 1.10.3) lib/mix/tasks/compile.all.ex:43: Mix.Tasks.Compile.All.with_logger_app/2
    (mix 1.10.3) lib/mix/task.ex:330: Mix.Task.run_task/3
    (mix 1.10.3) lib/mix/project.ex:352: Mix.Project.in_project/4
    (elixir 1.10.3) lib/file.ex:1544: File.cd!/2
    (mix 1.10.3) lib/mix/task.ex:430: anonymous fn/4 in Mix.Task.recur/1
    (elixir 1.10.3) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (mix 1.10.3) lib/mix/task.ex:429: Mix.Task.recur/1
    (mix 1.10.3) lib/mix/project_stack.ex:181: Mix.ProjectStack.recur/1
    (mix 1.10.3) lib/mix/tasks/compile.ex:96: Mix.Tasks.Compile.run/1
```

### Expected behavior

Compile without crash and proper output

**Comment by josevalim** [comment]
Can you please provide an application that reproduces the error? `--all-warnings` is working fine on the project I am working on, where a similar warning for `Supervisor.Spec.supervise/2` was added, but to be sure we need to be able to reproduce the exact warning you have. Thank you.

**Comment by mojidabckuu** [comment]
@josevalim Sure, https://github.com/mojidabckuu/all_warnings_crash Hope it helps

**In short**
```elixir
defmodule AllWarnSample.Supervisor do
  use Supervisor

  def start_link() do
    Supervisor.start_link(__MODULE__, [])
  end

  def init(_) do
    children = [
      worker(AllWarnSample.Worker, [])
    ]

    supervise(children, strategy: :one_for_one)
  end
end
```

Plus

```elixir
defmodule AllWarnSample.Worker do
  use GenServer

  def init(opts) do
    {:ok, opts}
  end

  def start_link(opts \\ []) do
    name = Keyword.get(opts, :name, __MODULE__)
    GenServer.start_link(__MODULE__, opts, name: name)
  end
end
```

If no children specified(empty list `[]`) then it compiles fine :)

**Comment by josevalim** [comment]
Thanks, fixed on master and v1.10. It will be out in the next release.
